src/main/java/com/faforever/client/map/generator/MapGeneratorService.java

```diff
@@ -0,0 +1,147 @@
+import java.io.IOException;
+import java.lang.invoke.MethodHandles;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Arrays;
+import java.util.Random;
+import java.util.concurrent.CompletableFuture;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import static com.github.nocatch.NoCatch.noCatch;
+
+@Lazy
+@Service
+public class MapGeneratorService {
+
+  private static final Logger logger = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+  @Getter
+  private static final String GENERATED_MAP_NAME = "NeroxisMapGenerator_%s_%d";
+  private static final String GENERATOR_DEFAULT_VERSION = "0.1.1";
+
+  @Getter
+  private static final String GENERATOR_EXECUTABLE_FILENAME = "MapGenerator_%s.jar";
+  @VisibleForTesting
+  public static final String GENERATOR_EXECUTABLE_SUB_DIRECTORY = "map_generator";
+  @Getter
+  private static final int GENERATION_TIMEOUT_SECONDS = 20;
+  private static final Pattern VERSION_PATTERN = Pattern.compile("\\d\\d?\\d?\\.\\d\\d?\\d?\\.\\d\\d?\\d?");
+  private static final Pattern GENERATED_MAP_PATTERN = Pattern.compile("NeroxisMapGenerator_(" + VERSION_PATTERN + ")_(-?\\d+)");
+  @Getter
+  private final Path generatorExecutablePath;
+  private final ApplicationContext applicationContext;
+  private final PreferencesService preferencesService;
+  private final TaskService taskService;
+
+  @Getter
+  private Path customMapsDirectory;
+  private Random seedGenerator;
+
+  @Inject
+  public MapGeneratorService(ApplicationContext applicationContext, PreferencesService preferencesService, TaskService taskService) {
+    this.applicationContext = applicationContext;
+    this.preferencesService = preferencesService;
+    this.taskService = taskService;
+
+    generatorExecutablePath = preferencesService.getFafDataDirectory().resolve(GENERATOR_EXECUTABLE_SUB_DIRECTORY);
+    if (!Files.exists(generatorExecutablePath)) {
+      try {
+        Files.createDirectory(generatorExecutablePath);
+      } catch (IOException e) {
+        logger.error("Could not create map generator executable directory.", e);
+      }
+    }
+
+    seedGenerator = new Random();
+
+    customMapsDirectory = this.preferencesService.getPreferences().getForgedAlliance().getCustomMapsDirectory();
+    deleteGeneratedMaps();
+  }
+
+
+  private void deleteGeneratedMaps() {
+    logger.info("Deleting leftover generated maps...");
+
+    if (customMapsDirectory != null && customMapsDirectory.toFile().exists()) {
+      Arrays.stream(customMapsDirectory.toFile().listFiles())
+          .filter(File::isDirectory)
+          .filter(f -> GENERATED_MAP_PATTERN.matcher(f.getName()).matches())
+          .map(File::toPath)
+          .forEach(f -> noCatch(() -> FileUtils.deleteRecursively(f)));
+    }
+  }
+
+
+  public CompletableFuture<String> generateMap() {
+    return generateMap(GENERATOR_DEFAULT_VERSION, seedGenerator.nextLong());
+  }
+
+
+  //TODO: PROTECT AGAINST mapName INJECTION!!!!!
```

Comment:
I dont kno what you mean there but if this is important fix it and at least clearify what you mean